### PR TITLE
OBX-23 CLIA fixes

### PR DESCRIPTION
**Pull request by RickHawesUSDS**
This PR fixes a bug in our HL7 message that was reported WA.

The OBX-23.11 component was being set to 'CLIA'. There isn't a OBX-23.11 component in the spec. 

## Changes
- Only set the Universal ID type to 'CLIA' for HDs and CE fields types
- Fixed and expanded the metadata about the HL7 fields

## Alternatives considered
I did consider a larger 

### Testing
- [x] Tested locally?
- [ ] Ran `quickTest all`?
- [x] Ran `./prime test` against local Docker ReportStream container?
- [ ] Downloaded a file from `http://localhost:7071/api/download`?
- [x] Added tests?

### Specific Security-related subjects a reviewer should pay specific attention to
No to all

- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Fixes
*List GitHub issues this PR fixes*
- #1580 

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

### CODE REVIEW

File: prime-router/settings/organizations.yml [hunk 1]
```diff
@@ -304,7 +304,21 @@
...
+        receivingApplicationOID: 2.16.840.1.114222.4.1.185.1
+        receivingFacilityName: VDH
+        receivingFacilityOID: 2.16.840.1.114222.4.1.185
+    - name: elr-secondary
+      organizationName: vt-doh
+      topic: covid-19
+      jurisdictionalFilter:
+        - "orEquals(ordering_facility_state, VT, patient_state, VT)"
+      reverseTheQualityFilter: true
       deidentify: false
```
Comment (RickHawesUSDS) [inline prime-router/settings/organizations.yml:322]: Had to VT settings locally to test VT translations

File: prime-router/src/main/kotlin/Translator.kt [hunk 2]
```diff
@@ -161,10 +161,10 @@ class Translator(private val metadata: Metadata, private val settings: SettingsP
...
     ): Pair<Report, Receiver>? {
         if (input.isEmpty()) return null
-        val service = settings.findReceiver(toReceiver) ?: error("invalid service name $toReceiver")
-        val mappedReport = translateByReceiver(input, service, defaultValues)
+        val receiver = settings.findReceiver(toReceiver) ?: error("invalid receiver name $toReceiver")
+        val mappedReport = translateByReceiver(input, receiver, defaultValues)
         if (mappedReport.itemCount == 0) return null
-        return Pair(mappedReport, service)
+        return Pair(mappedReport, receiver)
     }
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/Translator.kt:169]: Just a rename of service to receiver

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt [hunk 3]
```diff
@@ -433,7 +433,7 @@ class Hl7Serializer(val metadata: Metadata) : Logging {
                     val truncatedValue = if (
                         value.length > HD_TRUNCATION_LIMIT &&
                         element.type == Element.Type.TEXT &&
-                        hl7Field in HD_FIELDS &&
+                        hl7Field in HD_FIELDS_LOCAL &&
                         hl7Config?.truncateHDNamespaceIds == true
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:437]: Destinguished between Local HD and Universal HDs. Added a comment to explain.

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt [hunk 4]
```diff
@@ -642,6 +637,27 @@ class Hl7Serializer(val metadata: Metadata) : Logging {
...
+        when (hl7Field) {
+            in HD_FIELDS_UNIVERSAL,
+            in CE_FIELDS -> {
+                val nextComponent = nextComponent(pathSpec)
+                terser.set(nextComponent, "CLIA")
+            }
+        }
+    }
+
     /**
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:661]: New function replaces the old always set a CLIA type logic.

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt [hunk 5]
```diff
@@ -1103,6 +1119,32 @@ class Hl7Serializer(val metadata: Metadata) : Logging {
...
+         * List of fields that have the universal HD type
+         */
+        val HD_FIELDS_UNIVERSAL = listOf("MSH-4-2", "OBR-3-3", "OBR-2-3", "ORC-3-3", "ORC-2-3",
+            "PID-3-4-2", "PID-3-6-2", "SPM-2-1-3", "SPM-2-2-3")
+
+        /**
+         * List of fields that have a CE type
+         */
+        val CE_FIELDS = listOf("OBX-15-1")
     }
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:1149]: Same approch as before, but with little more metadata about the type of HL7 field.

File: prime-router/src/testIntegration/resources/datatests/CSV_to_HL7/sample-single-pdi-20210608-0002.hl7 [hunk 6]
```diff
@@ -5,7 +5,7 @@ SFT|Centers for Disease Control and Prevention|0.1-SNAPSHOT|PRIME Data Hub|0.1-S
...
 ORC|RE|885617^Any lab USA^36D1332559^CLIA|885617^Any lab USA^36D1332559^CLIA|||||||||8997885262^Hane^Cornelius||(285)138-9942^WPN^PH^^1^285^1389942|20210603||||||Any facility USA|7143 Wilford Meadow^^^IG|(235)675-4082^WPN^PH^^1^235^6754082|395 Tenesha Stravenue^^^IG
 OBR|1|885617^Any lab USA^36D1332559^CLIA|885617^Any lab USA^36D1332559^CLIA||||202106040725-0400|202106040725-0400||||||||8997885262^Hane^Cornelius|(285)138-9942^WPN^PH^^1^285^1389942|||||202105280000-0500|||C
-OBX|1|CWE|94558-4||840539006^Disease caused by sever acute respiratory syndrome coronavirus 2 (disorder)^SCT^^^^20200309|||A|||C|||202106040725-0400|36D1332559^CLIA||||202105310000-0500||||Any lab USA^^^^^^XX^^^36D1332559^CLIA|386 Mary Manors^^^IG
+OBX|1|CWE|94558-4||840539006^Disease caused by sever acute respiratory syndrome coronavirus 2 (disorder)^SCT^^^^20200309|||A|||C|||202106040725-0400|36D1332559^CLIA||||202105310000-0500||||Any lab USA^^^^^^XX^^^36D1332559|386 Mary Manors^^^IG
 OBX|2|DT|65222-2^Date and time of symptom onset^LN^^^^2.68||20210606||||||F|||202106040725-0400|36D1332559||||202105310000-0500||||Any lab USA^^^^^^XX^^^36D1332559|386 Mary Manors^^^IG|||||QST
```
Comment (RickHawesUSDS) [inline prime-router/src/testIntegration/resources/datatests/CSV_to_HL7/sample-single-pdi-20210608-0002.hl7:9]: This is the erroneous field.

**Comment by RickHawesUSDS** [comment]
Another test that I did was take a production VT result, deidentify it, and see the effect on the resultant HL7 message had unwanted changes.
 
VT HL7 message before:
```
FHS|^~\&|CDC PRIME - Atlanta, Georgia (Dekalb)^2.16.840.1.114222.4.1.237821^ISO|CDC PRIME - Atlanta, Georgia (Dekalb)^2.16.840.1.114222.4.1.237821^ISO|NBS^2.16.840.1.114222.4.1.185.1^ISO|VDH^2.16.840.1.114222.4.1.185^ISO|202107171704-0700
BHS|^~\&|CDC PRIME - Atlanta, Georgia (Dekalb)^2.16.840.1.114222.4.1.237821^ISO|CDC PRIME - Atlanta, Georgia (Dekalb)^2.16.840.1.114222.4.1.237821^ISO|NBS^2.16.840.1.114222.4.1.185.1^ISO|VDH^2.16.840.1.114222.4.1.185^ISO|202107171704-0700
MSH|^~\&|CDC PRIME - Atlanta, Georgia (Dekalb)^2.16.840.1.114222.4.1.237821^ISO|^n/a^CLIA|NBS^2.16.840.1.114222.4.1.185.1^ISO|VDH^2.16.840.1.114222.4.1.185^ISO|20210717170441.5024-0700||ORU^R01^ORU_R01|2021071622-dsr\S\2021071622-e8222070cc9e4a8697032d32ceff9d6b|P|2.5.1|||NE|NE|USA|UNICODE UTF-8|||PHLabReport-NoAck^ELR_Receiver^2.16.840.1.113883.9.11^ISO
SFT|Centers for Disease Control and Prevention|0.1-SNAPSHOT|PRIME Data Hub|0.1-SNAPSHOT||20210717
PID|1||60eac18c6ef3830034fb8092^^^&n/a&CLIA^^&n/a&CLIA||Fake^Joe^^^^^L||19680314|O||UNK^Unknown^NULLFL^^^^2.5.1|^^^VT^5146^USA^^^50025||^NET^Internet^fakename@gmail.com~(617)555-5555^PRN^PH^^1^617^5555555|||||||||U^Unknown^HL70189^^^^2.9||||||||N
ORC|RE|^^n/a^CLIA|^^n/a^CLIA|||||||||n/a^n/a^n/a^^^^^^CMS&2.16.840.1.113883.3.249&ISO^^^^NPI|||20210711|||||||||n/a^n/a^n/a^n/a
OBR|1|^^n/a^CLIA|^^n/a^CLIA|95409-9^SARS coronavirus 2 (COVID19) N gene [Presence] in Nose by NAA with probe detection^LN^^^^2.69|||202107110000-0500|202107110000-0500||||||||n/a^n/a^n/a^^^^^^CMS&2.16.840.1.113883.3.249&ISO^^^^NPI
OBX|1|CWE|95409-9^SARS coronavirus 2 (COVID19) N gene [Presence] in Nose by NAA with probe detection^LN^^^^2.69||260415000^Not detected^SCT|||N^Normal (applies to non-numeric results)^HL70078^^^^2.7||||||202107110000-0500|n/a^CLIA||^^^^^^2.69||202107110000-0500|||202107110000-0500|^^^^^CLIA&2.16.840.1.113883.19.4.6&ISO^XX^^^n/a^CLIA
OBX|2|CWE|30525-0^Age^LN^^^^2.68||53||||||F|||202107110000-0500|n/a||||202107110000-0500||||^^^^^CLIA&2.16.840.1.113883.19.4.6&ISO^XX^^^n/a||||||QST
SPM|1|2021071622-dsr\S\2021071622-e8222070cc9e4a8697032d32ceff9d6b&&n/a&CLIA^&&n/a&CLIA||445297001^Swab of internal nose^SCT^^^^2.67|||||||||||||202107110000-0500|20210711000006.0000-0500
BTS|1
FTS|1
```
And the message after the change. Notice that only the OBX-23-11 and timestamps change. 
```
FHS|^~\&|CDC PRIME - Atlanta, Georgia (Dekalb)^2.16.840.1.114222.4.1.237821^ISO|CDC PRIME - Atlanta, Georgia (Dekalb)^2.16.840.1.114222.4.1.237821^ISO|NBS^2.16.840.1.114222.4.1.185.1^ISO|VDH^2.16.840.1.114222.4.1.185^ISO|202107171714-0700
BHS|^~\&|CDC PRIME - Atlanta, Georgia (Dekalb)^2.16.840.1.114222.4.1.237821^ISO|CDC PRIME - Atlanta, Georgia (Dekalb)^2.16.840.1.114222.4.1.237821^ISO|NBS^2.16.840.1.114222.4.1.185.1^ISO|VDH^2.16.840.1.114222.4.1.185^ISO|202107171714-0700
MSH|^~\&|CDC PRIME - Atlanta, Georgia (Dekalb)^2.16.840.1.114222.4.1.237821^ISO|^n/a^CLIA|NBS^2.16.840.1.114222.4.1.185.1^ISO|VDH^2.16.840.1.114222.4.1.185^ISO|20210717171409.3432-0700||ORU^R01^ORU_R01|2021071622-dsr\S\2021071622-e8222070cc9e4a8697032d32ceff9d6b|P|2.5.1|||NE|NE|USA|UNICODE UTF-8|||PHLabReport-NoAck^ELR_Receiver^2.16.840.1.113883.9.11^ISO
SFT|Centers for Disease Control and Prevention|0.1-SNAPSHOT|PRIME Data Hub|0.1-SNAPSHOT||20210717
PID|1||60eac18c6ef3830034fb8092^^^&n/a&CLIA^^&n/a&CLIA||Fake^Joe^^^^^L||19680314|O||UNK^Unknown^NULLFL^^^^2.5.1|^^^VT^5146^USA^^^50025||^NET^Internet^fakename@gmail.com~(617)555-5555^PRN^PH^^1^617^5555555|||||||||U^Unknown^HL70189^^^^2.9||||||||N
ORC|RE|^^n/a^CLIA|^^n/a^CLIA|||||||||n/a^n/a^n/a^^^^^^CMS&2.16.840.1.113883.3.249&ISO^^^^NPI|||20210711|||||||||n/a^n/a^n/a^n/a
OBR|1|^^n/a^CLIA|^^n/a^CLIA|95409-9^SARS coronavirus 2 (COVID19) N gene [Presence] in Nose by NAA with probe detection^LN^^^^2.69|||202107110000-0500|202107110000-0500||||||||n/a^n/a^n/a^^^^^^CMS&2.16.840.1.113883.3.249&ISO^^^^NPI
OBX|1|CWE|95409-9^SARS coronavirus 2 (COVID19) N gene [Presence] in Nose by NAA with probe detection^LN^^^^2.69||260415000^Not detected^SCT|||N^Normal (applies to non-numeric results)^HL70078^^^^2.7||||||202107110000-0500|n/a^CLIA||^^^^^^2.69||202107110000-0500|||202107110000-0500|^^^^^CLIA&2.16.840.1.113883.19.4.6&ISO^XX^^^n/a
OBX|2|CWE|30525-0^Age^LN^^^^2.68||53||||||F|||202107110000-0500|n/a||||202107110000-0500||||^^^^^CLIA&2.16.840.1.113883.19.4.6&ISO^XX^^^n/a||||||QST
SPM|1|2021071622-dsr\S\2021071622-e8222070cc9e4a8697032d32ceff9d6b&&n/a&CLIA^&&n/a&CLIA||445297001^Swab of internal nose^SCT^^^^2.67|||||||||||||202107110000-0500|20210711000006.0000-0500
BTS|1
FTS|1
```

**Comment by carlosfelix2** [comment]
In your alternatives in the PR description, you just say "I did consider a larger".  Seems unfinished.

**Review by carlosfelix2 — APPROVED** [review]
You will need to also run 
`./prime test --run=compare` and check if you need data changes in those test files.  The data files are in src/main/resources

**Comment by RickHawesUSDS** [comment]
> You will need to also run
> `./prime test --run=compare` and check if you need data changes in those test files. The data files are in src/main/resources

Good point.

**Pull request merged**